### ISSUE TITLE
Pretraining loss explosion

### ISSUE THREAD

**Issue by mattgorb** [karpathy/nanoGPT#554]
I have been trying to get this repo working for several months, but my loss keeps exploding between 30k and 100k iterations.  

I have tried many things: 
Turn flash attention off ( based on this issue: https://github.com/karpathy/nanoGPT/issues/524)
Using fp16 (based in this: https://github.com/karpathy/nanoGPT/issues/468)
Using GPT-4 tokenizer (based on https://github.com/karpathy/nanoGPT/issues/468)

I have also tinkered with other setting such as gradient clipping, learning rate, etc. I keep my configuration at roughly 500k batch size.  

I am lost on what to try next.  Did anyone else fix this issue?  

I have gotten GPT-2 Small down to about 3.0 loss.

**Comment by beeeai** [comment]
@mattgorb @HarrisonUnifyAI I have not changed the default settings, other than the seed.
I get this explosion which starts around 30k as well. Do you think flash attention is the reason?  
(The purple one is all default values, including the seed set to 1337. The blue one is another seed.)
![image](https://github.com/user-attachments/assets/1f7b749e-37c5-4ed8-86f9-b181b44fb19e)

**Comment by pyotr777** [comment]
I had similar results for training on 8xA100-SXM4-40GB. The validation loss graph is below. I changed the micro batch size to fit into GPU memory 64 → 16. To keep the total batch size, I increased accumulation steps:

```
total desired batch size: 524288
=> calculated gradient accumulation steps: 4

```
I ran training three times with the same parameters. Training collapsed all three times, but the timing was different.
(I am trying to attach images, but not sure if it works.)

<!-- Uploading "newplot-185.jpg"... —>

<!-- Uploading "newplot-185.png"... —>

**Comment by abs-hello** [comment]
Hi, I'm trying to reproduce this issue for a deeper numerical analysis. Could you share the exact training configuration (hyperparameters, etc.)?
Also, having access to the full checkpoint where this behavior appears would be extremely valuable.
Thanks!

**Comment by abs-hello** [comment]
> My loss also exploded on an initial attempt with 8 A100 instance. I disabled flash attention and that seems to have fixed it, though it is much slower.

Could you please show your configurations which are different from the original project.

**Comment by abs-hello** [comment]
> I had similar results for training on 8xA100-SXM4-40GB. The validation loss graph is below. I changed the micro batch size to fit into GPU memory 64 → 16. To keep the total batch size, I increased accumulation steps:
> 
> ```
> total desired batch size: 524288
> => calculated gradient accumulation steps: 4
> ```
> 
> I ran training **three times** with the same parameters. Training collapsed all three times, but the timing was different. (Attaching an image didn't work for me in Safari, but worked in Chrome browser.)
> 
> ![Image](https://github.com/user-attachments/assets/ac82394a-2588-4f36-a9ea-990435df919b)

Did you change the seed or any other parameters?

**Comment by pyotr777** [comment]
> Did you change the seed or any other parameters?

No. I used the original script, just changed the microbatch size.  But I don't remember any details now.
